README.md
# weewx-rainrate
*Open source plugin for WeeWX software.

Copyright (C)2022-2023 by John A Kline (john@johnkline.com)

**This extension requires Python 3.7 or later and WeeWX 4.**


## Description

weewx-rainrate is a WeeWX service that attempts to produce a
"better" rainRate in loop packets (and archive records) for
siphon tipping bucket rain gauges.

This extension will be useful for tipping
rain gauges that use a siphon for better accuracy over a wide
range of rainfall.  These professional gauges maintain their
accuracy over a wide range of rain intensity, but are
unsuitable for computing rain rate via the time
between two tips.  The reason for the unsuitability is that
a single discharge of the siphon may result in multiple tips
(in close sucession).  The result of two tips in close
succession will be a wildly overstated rain rate.

The impetus for this extension was the author's purchase of a
professional HyQuest Solutions TB3 tipping rain gauge with
siphon.  It is accurate to 2% at any rain intensity, but with
the siphon, two tips can come in quick succession.

The extension was tested with a HyQuest Solutions TB3 siphon
tipping bucket rain gauge and using a HyQuest Solutions TB7 (non-siphon)
tipping bucket rain gauge as a reference (for rain rate).

## Visualization

The following two images show rain rates during two storms in Palo Alto in
December of 2022.  The first yielded 0.68" of rain on December 1.
The second is the first part of a storm on December 3 (measuring 0.32" of rain).

The blue lines in each graph are the untouched rates of a HyQuest TB7 tipping bucket, which does not have a siphon and, as such, is provided as a reference.

The red lines show the rain rate from a HyQuest TB3 tipping bucket, which does have a siphon and, as such, can suffer from multiple tips at once.
The left graph in each image shows the untouched rain rate of the TB3.  As one
can see, nonsensical rates of up to 16" per hour are shown.

On the right, this rainrate extension is used. The TB3 shows rain rates that track reasonably well with the TB7.

![Dec 1, 2022 storm](Dec1BeforeAndAfter.png)
Dec 1, 2022 storm.  Reference TB7 in blue.  TB3 (red) on left without extension, on right with this extension.

![Dec 3, 2022 storm](Dec3BeforeAndAfter.png)
Dec 3, 2022 storm.  Reference TB7 in blue.  TB3 (red) on left without extension, on right with this extension.

The TB3 vs. TB7 `rain` and `rainRate` values in  archive records for part of a rain storm
on the night of Jan 4/5, 2023 can be found [here](TB3vsTB7_2023-01-04_05.md).

## Algorithm

The key feature of this extension is that, when multiple tips are encountered in a loop record, the extension (for the purposes of calculating rain rate) records 1/2 of the amount at the current time.  It records the other half of the amount at a point midway between the last rain reported and the current time.

The exceptional case is where a multi-tip reading is encountered on the first tip of a storm (where a storm is defined as the time of the first tip through to a period of 30 minutes with no tips).  For this case, 0.01 of rain is recorded at the current time and remainder is recorded 15m earlier.

The rain rate is determined by choosing the *lower* of the following two calculations:

1. The rate between the last two tips:
`3600.0 * rain_amount_of_most_recent_tip / (time_of_last_tip - time_of_previous_tip)`
The `rain_amount_of_most_recent_tip` will be 0.01 (since
multi-tips are split up).  The `time_of_previous_tip` is not the last rain
event; rather, it's the one before that.

1. The rate between the last tip and now (as if a new tip occurred now).
`3600.0 * 0.01 / (now - time_of_most_recent_tip)`  Note: if the curent loop packet
contains a tip, this second rate is not calculated.

# Installation Instructions

1. Download the lastest release, weewx-rainrate-0.32.zip, from the
   [GitHub Repository](https://github.com/chaunceygardiner/weewx-rainrate).

1. Run the following command.

   `sudo /home/weewx/bin/wee_extension --install weewx-rainrate-0.32.zip`

   Note: this command assumes weewx is installed in /home/weewx.  If it's installed
   elsewhere, adjust the path of wee_extension accordingly.

1. Restart WeeWX.

1. The following entry is created in `weewx.conf`.  To disable `weewx-rainrate` without
   uninstalling it, change the enable line to false.
```
[RainRate]
    enable = true
```

## Why require Python 3.7 or later?

weewx-rainrate code includes type annotation which do not work with Python 2, nor in
earlier versions of Python 3.

## Licensing

weewx-rainrate is licensed under the GNU Public License v3.


changes.txt
weewx-rainrate change history
-----------------------------

0.32 Release 2023/01/?? 
-----------------------
When calculated rainRate falls below 0.04, report it as 0.0.

0.31 Release 2023/01/?? 
-----------------------
Although never observed, handle case where double tip spans two loop records.
Although never observed, handle case where more than two tips in a loop record.

0.30 Release 2023/01/??
-----------------------
A rainrate, once computed, will stick until a point arrives in time where
that rainrate cannot be acheived because, even if the next tip were to
occur, the rainrate would be lower than the present rate.  At that point,
continue dropping the rain rate to a rate that could be achieved if a
tip were to occur.  Continue until another tip or until a rate of 0.0
is achieved (because the previous tips fell off the extenstions tracking).

0.20 Release 2022/12/31
-----------------------
A tweak to the decay of rainRate.  Rain rate is computed as the average of:
rainRate1: 3600.0 * rain_amount_of_most_recent_tip / (time_of_last_tip - time_of_previous_tip)
rainRate2: 3600.0 * rain_amount_of_most_recent_tip / (now - time_of_previous_rain_event)
As time passes, rainRate2 becomes more prominent (y=x^sqrt(2))

0.19 Release 2022/12/27
-----------------------
Decay the rainRate more slowly as time passes from the last tip (See
README) to more closely match the vanatage console's rain rate decay.

0.18 Release 2022/12/27
-----------------------
Bug fix, was decrementing rain in pkt when multi-tips occurred.

0.17 Release 2022/12/12
-----------------------
Fix errors mistakenly released in 0.16.

0.16 Release 2022/12/05
-----------------------
1. Completely revamp calculations.  Now the calculation is
   3600 * 0.01 / (now - the time of two tips ago)
   0.02 readings of rain are split with half now and half
   recorded (for purposes of calculating rainrate) as taking
   place halfway in between now and the previous tip.

2. Archive record now reports highest rain rate of the
   archive period.


0.15 Release 2022/12/04
-----------------------
Prior versions of this extension update rain rates in loop packets, but
left archive records untouched.  Archive record rain rates are now
updated to be the average of the rain rate in the loop records for
that archive period.  If there are no loop record rain rates (most likely
because weewx is doing a catchup at startup), the rain rate will be
computed as the total rainfall for this archive period averaged over 15m.

0.14 Release 2022/12/02
-----------------------
Now tested against a significant storm and against a quality non-siphon
tipping bucket (HyQuest Solutions TB7).  Tweaked so that the TB3 (siphone)
resembles the TB7.  Changes are: only use 15 average if 0.01 or 0.02 fell
within last 15m.  Only consider 10 through 15m buckets when 0.03 has
fallen in the last 15m.

0.13 Release 2022/11/18
-----------------------
Min timespan to average now 4m (was 3m).
Buckets are now (back to) 1m (rather than 30s).
Both of these changes were made after considering the paper,
Estimating Rain Rates from Tipping-Bucket Rain Gauge Measurements,
at https://ntrs.nasa.gov/api/citations/20070016690/downloads/20070016690.pdf
"One-minute rain rates suffer substantial errors, especially at low rain rates.
When one-minute rain rates are averaged to 4 -7 minute scales, the errors
dramatically reduce."

0.12 Release 2022/11/13
-----------------------
Min timespan to average now 3m (was 2m).

0.11 Release 2022/11/11
-----------------------
1. Move to 30s increments, from 1 min. increments, for more responsive rain rate.

0.1 Release 2022/11/10
----------------------
1. Never use the last 60s for rainRate as it is thought to be too noisy.
2. Refactor for testing.
3. Add extensive tests.

0.02 Release 2022/11/09
-----------------------
1. Special case the algorithm for rain rate when low volumes of rain are observed.

0.01 Release 2022/11/08
-----------------------
1. Initial release.


install.py
# Copyright 2022 by John A Kline <john@johnkline.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import sys
import weewx
from setup import ExtensionInstaller

def loader():
    if sys.version_info[0] < 3 or (sys.version_info[0] == 3 and sys.version_info[1] < 7):
        sys.exit("weewx-rainrate requires Python 3.7 or later, found %s.%s" % (
            sys.version_info[0], sys.version_info[1]))

    if weewx.__version__ < "4":
        sys.exit("weewx-rainrate requires WeeWX 4, found %s" % weewx.__version__)

    return RainRateInstaller()

class RainRateInstaller(ExtensionInstaller):
    def __init__(self):
        super(RainRateInstaller, self).__init__(
            version = "0.31",
            name = 'rainrate',
            description = 'Inserts/updates rainRate observations in loop packets.',
            author = "John A Kline",
            author_email = "john@johnkline.com",
            data_services = 'user.rainrate.RainRate',
            config = {
                'RainRate': {
                    'enable'     : 'true',
                },
            },
            files = [
                ('bin/user', [
                    'bin/user/rainrate.py',
                    ]),
            ])
